### PR TITLE
feat: support Let's Encrypt tls.

### PR DESCRIPTION
**Pull request by appleboy**
Support Let's Encrypt for a Go server application.

cc @tboerger @javierprovecho

### CODE REVIEW

**Comment by nazwa** [comment]
Hi @appleboy! I had no idea how simple this could be, nice work.

Do you know if it regenerates the certificate automatically when the app is running for a long time? If i don't restart the server for 3 months, will it swap the cert on the fly or will it continue serving the expired one? (I know nothing about autocert)

File: gin.go [hunk 1]
```diff
@@ -255,6 +257,35 @@ func (engine *Engine) RunTLS(addr string, certFile string, keyFile string) (err
...
+	}
+
+	//your domain here
+	if len(domain) != 0 {
+		m.HostPolicy = autocert.HostWhitelist(domain...)
+	}
+
+	// folder for storing certificates
+	if cache != "" {
+		m.Cache = autocert.DirCache(cache)
```
Comment (tboerger) [inline gin.go]: A hard-coded file cache is not so cool

**Comment by tboerger** [comment]
I would prefer an example how to integrate that instead of having a hard dependency, especially because others prefer Lego.

**Comment by tboerger** [comment]
And I forgot to mention, AFAIK let's encrypt enforces port 443, so the port config is more or less useless

File: gin.go (hunk 1 above)
Comment (appleboy) [inline gin.go]: @tboerger This is the cache folder path, not file.

**Comment by appleboy** [comment]
@tboerger See example: https://github.com/gin-gonic/gin/pull/863/files#diff-c00dc46ee4a20e12b779d7a4469af81a `examples/auto-tls/main.go`

**Comment by tboerger** [comment]
I'm talking about an example only without adding it to the core

**Comment by appleboy** [comment]
@tboerger I got it.

**Comment by appleboy** [comment]
@tboerger @nazwa 

Updated. Please review again.

* remove hard-coded file cache
* only listen on 443 port
* update example code

**Comment by appleboy** [comment]
Notes: This feature only works for 1.7 version or higher.

**Comment by tboerger** [comment]
That way the core still got a hard requirement for the autocert lib :(

**Comment by appleboy** [comment]
@tboerger Updated. Just add an example of how to integrate with Let's Encrypt service.

**Comment by appleboy** [comment]
@tboerger Could you explain why we can't use `golang.org/x/crypto/acme/autocert` ?

**Comment by tboerger** [comment]
As I already said this is another dependency of the core with little benefit because it's so easy to integrate

**Comment by appleboy** [comment]
example code if we don't merge this PR.

```go
package main

import (
	"crypto/tls"
	"net/http"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/acme/autocert"
)

func main() {
	r := gin.Default()

	// Ping handler
	r.GET("/ping", func(c *gin.Context) {
		c.String(200, "pong")
	})

	m := autocert.Manager{
		Prompt:     autocert.AcceptTOS,
		HostPolicy: autocert.HostWhitelist("example1.com", "example2.com"),
		Cache:      autocert.DirCache("/var/www/.cache"),
	}

	s := &http.Server{
		Addr:      ":443",
		TLSConfig: &tls.Config{GetCertificate: m.GetCertificate},
		Handler:   r,
	}
	s.ListenAndServeTLS("", "")
}
```

the following example if merged PR

```go
package main

import (
	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/acme/autocert"
)

func main() {
	r := gin.Default()

	//folder for storing certificates
	gin.AutoTLSManager.Cache = autocert.DirCache("/var/www/.cache")

	// Ping handler
	r.GET("/ping", func(c *gin.Context) {
		c.String(200, "pong")
	})

	// Listen and Server in 0.0.0.0:443
	r.RunAutoTLS("example1.com", "example2.com")
}
```

I think it would be better if gin support `RunAutoTLS` function.

**Comment by tboerger** [comment]
:+1: for the first example

**Comment by javierprovecho** [comment]
@appleboy @tboerger what about a subpackage like binding or render not directly linked to gin main package? Example could be a mix of the two above:

```go
package main

import (
	"crypto/tls"
	"net/http"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/autotls"
)

func main() {
	r := gin.Default()

	// Ping handler
	r.GET("/ping", func(c *gin.Context) {
		c.String(200, "pong")
	})

	m := autocert.Manager{
		Prompt:     autocert.AcceptTOS,
		HostPolicy: autocert.HostWhitelist("example1.com", "example2.com"),
		Cache:      autocert.DirCache("/var/www/.cache"),
	}

	autotls.Run(r, autocert.AcceptTOS, "/var/www/.cache", "example1.com", "example2.com")
	autotls.RunWithManager(r, m)
}
```

**Comment by appleboy** [comment]
@javierprovecho autocert support 1-line LetsEncrypt HTTPS servers.

https://github.com/golang/crypto/commit/b020702ab212964a017cbb8f7db52b5367017a4d

example:

```go
package main
 
import (
    "log"
    "net/http"
 
    "github.com/gin-gonic/gin"
    "golang.org/x/crypto/acme/autocert"
)
 
func main() {
    r := gin.Default()
 
    // Ping handler
    r.GET("/ping", func(c *gin.Context) {
        c.String(200, "pong")
    })
 
    log.Fatal(http.Serve(autocert.NewListener("example1.com", "example2.com"), r))
}
```

so I update your example:

```diff
package main

import (
	"crypto/tls"
	"net/http"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/acme/autocert"
	"github.com/gin-gonic/gin/autotls"
)

func main() {
	r := gin.Default()

	// Ping handler
	r.GET("/ping", func(c *gin.Context) {
		c.String(200, "pong")
	})

	m := autocert.Manager{
		Prompt:     autocert.AcceptTOS,
		HostPolicy: autocert.HostWhitelist("example1.com", "example2.com"),
		Cache:      autocert.DirCache("/var/www/.cache"),
	}

-       autotls.Run(r, autocert.AcceptTOS, "/var/www/.cache", "example1.com", "example2.com")
+	autotls.Run(r, "example1.com", "example2.com")
	autotls.RunWithManager(r, m)
}
```

**Comment by javierprovecho** [comment]
@appleboy that would be the way to go IMO :smiley:

**Comment by appleboy** [comment]
@javierprovecho OK. I will create sub package which name is `autotls`

**Comment by appleboy** [comment]
@javierprovecho I think it would be the better way that creates another repo as sub package like https://github.com/gin-gonic/autotls and just make some example in gin repo.

cc @tboerger

**Comment by appleboy** [comment]
@javierprovecho I don’t have permission to create new repo in https://github.com/gin-gonic

**Comment by appleboy** [comment]
@tboerger @javierprovecho Done. Just add some examples for Let's Encrypt

**Comment by appleboy** [comment]
Also see the autotls package: https://github.com/gin-gonic/autotls

**Pull request merged**